In [1]:
import os
import numpy as np
import pandas as pd


df = pd.read_csv('/kaggle/input/geeksforgeeks-articles/articles.csv')
df.dropna(inplace = True)

In [2]:
df

title         author_id  \
0                 5 Best Practices For Writing SQL Joins       priyankab14   
1                           Foundation CSS Dropdown Menu  ishankhandelwals   
2           Top 20 Excel Shortcuts That You Need To Know       priyankab14   
3                              Servlet – Fetching Result   nishatiwari1719   
4                                       Suffix Sum Array          rohit768   
...                                                  ...               ...   
34569              Data Structures | Queue | Question 11     GeeksforGeeks   
34570        Data Structures | Binary Trees | Question 1     GeeksforGeeks   
34571                           Amazon Interview | Set 9     GeeksforGeeks   
34572  Python Program for Rat in a Maze | Backtracking-2     GeeksforGeeks   
34573            Data Structures and Algorithms | Set 21     GeeksforGeeks   

       last_updated                                               link  \
0      21 Feb, 2022  https://www.geeksforgeeks.org/5-best-practices...   
1      20 Feb, 2022  https://www.geeksforgeeks.org/foundation-css-d...   
2      17 Feb, 2022  https://www.geeksforgeeks.org/top-20-excel-sho...   
3      17 Feb, 2022  https://www.geeksforgeeks.org/servlet-fetching...   
4      21 Feb, 2022    https://www.geeksforgeeks.org/suffix-sum-array/   
...             ...                                                ...   
34569  28 Jun, 2021  https://www.geeksforgeeks.org/data-structures-...   
34570  28 Jun, 2021  https://www.geeksforgeeks.org/data-structures-...   
34571  28 Apr, 2017  https://www.geeksforgeeks.org/amazon-interview...   
34572  02 Aug, 2021  https://www.geeksforgeeks.org/python-program-f...   
34573  27 Mar, 2017  https://www.geeksforgeeks.org/data-structures-...   

      category  
0         easy  
1         easy  
2         easy  
3         easy  
4         easy  
...        ...  
34569   expert  
34570   expert  
34571   expert  
34572   expert  
34573   expert  

[34551 rows x 5 columns]

In [3]:
df.isnull().sum()

title           0
author_id       0
last_updated    0
link            0
category        0
dtype: int64

## 1) How many authors do we have in the dataset?

In [4]:
df['author_id'].describe()

count             34551
unique             5589
top       GeeksforGeeks
freq              11957
Name: author_id, dtype: object

## unique             5589


## 2) How many easy articles are publised on GeeksforGeeks?

In [5]:
len(df[df['category'] == 'easy'])

9654

## 9654


## 3) Which category is having most number of articles?

In [6]:
df.groupby('category').size().sort_values(ascending = False)

category
medium    10431
easy       9654
basic      8140
hard       4232
expert     2094
dtype: int64

## medium    10431


## 4) What percent of articles are in expert category out of total number of articles?

In [7]:
round((len(df[df['category'] == 'expert'])/len(df))*100)

6

## 6%

## 5) Who has written most number of articles after GeeksforGeeks?

In [8]:
df.groupby('author_id').size().sort_values(ascending = False).head()

author_id
GeeksforGeeks    11957
ManasChhabra2      317
Striver            265
manjeet_04         246
Chinmoy Lenka      192
dtype: int64

## ManasChhabra2      317


## 6) Who has written most number of articles in Expert Category after GeeksforGeeks?

In [9]:
df[df['category'] == 'expert'].groupby('author_id').size().sort_values(ascending = False).head()

author_id
GeeksforGeeks      285
mishrapriyank17     36
pintusaini          34
harkiran78          24
zack_aayush         22
dtype: int64

## mishrapriyank17     36


In [10]:
df

title         author_id  \
0                 5 Best Practices For Writing SQL Joins       priyankab14   
1                           Foundation CSS Dropdown Menu  ishankhandelwals   
2           Top 20 Excel Shortcuts That You Need To Know       priyankab14   
3                              Servlet – Fetching Result   nishatiwari1719   
4                                       Suffix Sum Array          rohit768   
...                                                  ...               ...   
34569              Data Structures | Queue | Question 11     GeeksforGeeks   
34570        Data Structures | Binary Trees | Question 1     GeeksforGeeks   
34571                           Amazon Interview | Set 9     GeeksforGeeks   
34572  Python Program for Rat in a Maze | Backtracking-2     GeeksforGeeks   
34573            Data Structures and Algorithms | Set 21     GeeksforGeeks   

       last_updated                                               link  \
0      21 Feb, 2022  https://www.geeksforgeeks.org/5-best-practices...   
1      20 Feb, 2022  https://www.geeksforgeeks.org/foundation-css-d...   
2      17 Feb, 2022  https://www.geeksforgeeks.org/top-20-excel-sho...   
3      17 Feb, 2022  https://www.geeksforgeeks.org/servlet-fetching...   
4      21 Feb, 2022    https://www.geeksforgeeks.org/suffix-sum-array/   
...             ...                                                ...   
34569  28 Jun, 2021  https://www.geeksforgeeks.org/data-structures-...   
34570  28 Jun, 2021  https://www.geeksforgeeks.org/data-structures-...   
34571  28 Apr, 2017  https://www.geeksforgeeks.org/amazon-interview...   
34572  02 Aug, 2021  https://www.geeksforgeeks.org/python-program-f...   
34573  27 Mar, 2017  https://www.geeksforgeeks.org/data-structures-...   

      category  
0         easy  
1         easy  
2         easy  
3         easy  
4         easy  
...        ...  
34569   expert  
34570   expert  
34571   expert  
34572   expert  
34573   expert  

[34551 rows x 5 columns]

In [11]:
df['last_updated'][0]

'21 Feb, 2022'

In [12]:
for date in df['last_updated']:
    if len(date)<12:
        df.drop(df[df['last_updated'] == date].index, inplace=True)


In [13]:
day = []
month = []
year = []
data = df.values
for date in data:
    day.append(date[2].split(" ")[0])
    month.append(date[2].split(" ")[1][:-1])
    year.append(date[2].split(" ")[2])

In [14]:
df['day'] = day
df['month'] = month
df['year'] = year

In [15]:
df['month']

0        Feb
1        Feb
2        Feb
3        Feb
4        Feb
        ... 
34569    Jun
34570    Jun
34571    Apr
34572    Aug
34573    Mar
Name: month, Length: 34455, dtype: object

## 7) Which day of the month has most number of articles published?

In [16]:
df.head()

title         author_id  \
0        5 Best Practices For Writing SQL Joins       priyankab14   
1                  Foundation CSS Dropdown Menu  ishankhandelwals   
2  Top 20 Excel Shortcuts That You Need To Know       priyankab14   
3                     Servlet – Fetching Result   nishatiwari1719   
4                              Suffix Sum Array          rohit768   

   last_updated                                               link category  \
0  21 Feb, 2022  https://www.geeksforgeeks.org/5-best-practices...     easy   
1  20 Feb, 2022  https://www.geeksforgeeks.org/foundation-css-d...     easy   
2  17 Feb, 2022  https://www.geeksforgeeks.org/top-20-excel-sho...     easy   
3  17 Feb, 2022  https://www.geeksforgeeks.org/servlet-fetching...     easy   
4  21 Feb, 2022    https://www.geeksforgeeks.org/suffix-sum-array/     easy   

  day month  year  
0  21   Feb  2022  
1  20   Feb  2022  
2  17   Feb  2022  
3  17   Feb  2022  
4  21   Feb  2022

In [17]:
df.groupby('day').size().sort_values(ascending = False)

day
28    2433
22    1315
21    1293
07    1186
10    1184
26    1175
17    1175
24    1175
11    1164
30    1149
29    1135
06    1122
19    1108
27    1105
18    1071
09    1070
14    1046
13    1020
23    1020
08    1012
03    1003
12     999
16     988
01     986
20     980
25     975
05     968
02     960
04     951
15     949
31     738
dtype: int64

## 28 day


## 8) Which day of July has most number of articles published?

In [18]:
df.head()

title         author_id  \
0        5 Best Practices For Writing SQL Joins       priyankab14   
1                  Foundation CSS Dropdown Menu  ishankhandelwals   
2  Top 20 Excel Shortcuts That You Need To Know       priyankab14   
3                     Servlet – Fetching Result   nishatiwari1719   
4                              Suffix Sum Array          rohit768   

   last_updated                                               link category  \
0  21 Feb, 2022  https://www.geeksforgeeks.org/5-best-practices...     easy   
1  20 Feb, 2022  https://www.geeksforgeeks.org/foundation-css-d...     easy   
2  17 Feb, 2022  https://www.geeksforgeeks.org/top-20-excel-sho...     easy   
3  17 Feb, 2022  https://www.geeksforgeeks.org/servlet-fetching...     easy   
4  21 Feb, 2022    https://www.geeksforgeeks.org/suffix-sum-array/     easy   

  day month  year  
0  21   Feb  2022  
1  20   Feb  2022  
2  17   Feb  2022  
3  17   Feb  2022  
4  21   Feb  2022

In [19]:
df[df['month'] == 'Jul'].groupby('day').size().sort_values(ascending = False)

day
07    166
06    146
16    129
30    126
02    117
09    113
26    113
23    107
01    107
19    104
22    103
15     99
28     99
29     98
08     97
10     96
12     91
05     91
13     76
20     74
21     72
31     69
17     68
04     65
27     63
03     63
11     62
18     59
14     54
25     51
24     45
dtype: int64

## 07 day of July


## 9) Which month has least number of articles published?

In [20]:
df.head()

title         author_id  \
0        5 Best Practices For Writing SQL Joins       priyankab14   
1                  Foundation CSS Dropdown Menu  ishankhandelwals   
2  Top 20 Excel Shortcuts That You Need To Know       priyankab14   
3                     Servlet – Fetching Result   nishatiwari1719   
4                              Suffix Sum Array          rohit768   

   last_updated                                               link category  \
0  21 Feb, 2022  https://www.geeksforgeeks.org/5-best-practices...     easy   
1  20 Feb, 2022  https://www.geeksforgeeks.org/foundation-css-d...     easy   
2  17 Feb, 2022  https://www.geeksforgeeks.org/top-20-excel-sho...     easy   
3  17 Feb, 2022  https://www.geeksforgeeks.org/servlet-fetching...     easy   
4  21 Feb, 2022    https://www.geeksforgeeks.org/suffix-sum-array/     easy   

  day month  year  
0  21   Feb  2022  
1  20   Feb  2022  
2  17   Feb  2022  
3  17   Feb  2022  
4  21   Feb  2022

In [21]:
df.groupby('month').size().sort_values()

month
Mar    1510
Apr    2318
Dec    2674
Feb    2716
Oct    2775
Jul    2823
Sep    2918
Nov    2932
Aug    3192
Jan    3206
May    3331
Jun    4060
dtype: int64

## march

## 10) Growth Rate of articles written 2021 as compare to 2020?

## Article written in 2020

In [22]:
l1 = len(df[df['year'] == '2020'])
l2 = len(df[df['year'] == '2021'])
print("2020::",l1)
print("2021::",l2)

2020:: 4625
2021:: 18613


## Article written in 2021

In [23]:
round((l2/l1)*100)

402

402 % is the growth rate